relay-cardinality/src/limiter.rs

```diff
@@ -434,4 +439,89 @@ mod tests {
             ]
         );
     }
+
+    #[test]
+    fn test_limiter_passive() {
+        struct RejectLimits;
+
+        impl Limiter for RejectLimits {
+            fn check_cardinality_limits<'a, I, T>(
+                &self,
+                _scoping: Scoping,
+                limits: &'a [CardinalityLimit],
+                entries: I,
+                rejections: &mut T,
+            ) -> Result<()>
+            where
+                I: IntoIterator<Item = Entry>,
+                T: Rejections<'a>,
+            {
+                for entry in entries {
+                    rejections.reject(&limits[entry.id.0 % limits.len()], entry.id);
+                }
+                Ok(())
+            }
+        }
+
+        let limiter = CardinalityLimiter::new(RejectLimits);
+        let limits = &[
+            CardinalityLimit {
+                id: "limit_passive".to_owned(),
+                passive: false,
+                window: SlidingWindow {
+                    window_seconds: 3600,
+                    granularity_seconds: 360,
+                },
+                limit: 10_000,
+                scope: CardinalityScope::Organization,
+                namespace: None,
+            },
+            CardinalityLimit {
+                id: "limit_enforced".to_owned(),
+                passive: true,
+                window: SlidingWindow {
+                    window_seconds: 3600,
+                    granularity_seconds: 360,
+                },
+                limit: 10_000,
+                scope: CardinalityScope::Organization,
+                namespace: None,
+            },
+        ];
+
+        let items = vec![
+            Item::new(0, MetricNamespace::Custom),
+            Item::new(1, MetricNamespace::Custom),
+            Item::new(2, MetricNamespace::Custom),
+            Item::new(3, MetricNamespace::Custom),
+            Item::new(4, MetricNamespace::Custom),
+            Item::new(5, MetricNamespace::Custom),
+        ];
+        let limited = limiter
+            .check_cardinality_limits(build_scoping(), limits, items)
+            .unwrap();
+
+        assert!(limited.has_rejections());
+        assert_eq!(limited.limits(), &limits.iter().collect());
+
+        // All passive items and no enforced (passive = False) should be accepted.
+        let rejected = limited.rejected().collect::<HashSet<_>>();
+        assert_eq!(
+            rejected,
+            HashSet::from([
+                &Item::new(0, MetricNamespace::Custom),
+                &Item::new(2, MetricNamespace::Custom),
+                &Item::new(4, MetricNamespace::Custom),
+            ])
+        );
+        drop(rejected); // NLL are broken here without the explicit drop
```

Comment:
What's NLL?

Reply:
Non lexical lifetimes.

The compiler should be smart enough to see that rejected is no longer used, but without the drop I cannot call `into_accepted` because there is still the reference to `rejected`.